lib/storage/metricnamestats/tracker.go

```diff
@@ -0,0 +1,572 @@
+}
+
+// MustLoadFrom inits tracker from the given on-disk path
+func MustLoadFrom(loadPath string, maxSizeBytes uint64) *Tracker {
+	mt, err := loadFrom(loadPath, maxSizeBytes)
+	if err != nil {
+		logger.Fatalf("unexpected error at tracker state load from path=%q: %s", loadPath, err)
+	}
+	return mt
+}
+
+func loadFrom(loadPath string, maxSizeBytes uint64) (*Tracker, error) {
+	mt := &Tracker{
+		maxSizeBytes:   maxSizeBytes,
+		store:          make(map[statKey]*statItem),
+		metricNamesBuf: make([]byte, 0, metricNameBufSize),
+		statItemBuf:    make([]statItem, 0, statItemBufSize),
+		cachePath:      loadPath,
+		getCurrentTs:   fasttime.UnixTimestamp,
+	}
+	mt.creationTs.Store(fasttime.UnixTimestamp())
+
+	f, err := os.Open(loadPath)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, fmt.Errorf("cannot access file content: %w", err)
+	}
+	defer f.Close()
+	if f != nil {
+		zr, err := gzip.NewReader(f)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create new gzip reader: %w", err)
+		}
+		reader := json.NewDecoder(zr)
+		var storedMaxSizeBytes uint64
+		if err := reader.Decode(&storedMaxSizeBytes); err != nil {
+			if errors.Is(err, io.EOF) {
+				return mt, nil
+			}
+			return nil, fmt.Errorf("cannot parse maxSizeBytes: %w", err)
+		}
+		if storedMaxSizeBytes > maxSizeBytes {
+			logger.Infof("Reseting tracker state due to changed maxSizeBytes from %d to %d.", storedMaxSizeBytes, maxSizeBytes)
+			return mt, nil
+		}
+		var creationTs uint64
+		if err := reader.Decode(&creationTs); err != nil {
+			return nil, fmt.Errorf("cannot parse creation timestamp: %w", err)
+		}
+		mt.creationTs.Store(creationTs)
+		var cnt uint64
+		var size uint64
+		var r recordForStore
+		for {
+			if err := reader.Decode(&r); err != nil {
+				if errors.Is(err, io.EOF) {
+					break
+				}
+				return nil, fmt.Errorf("cannot parse state record: %w", err)
+			}
+			si := mt.nextRecordLocked()
+
+			si.lastRequestTs.Store(r.LastRequestTs)
+			si.requestsCount.Store(r.RequestsCount)
+
+			metricName := bytesutil.ToUnsafeString(mt.allocateMetricNameLocked([]byte(r.MetricName)))
+			key := statKey{
+				projectID:  r.ProjectID,
+				accountID:  r.AccountID,
+				metricName: metricName,
+			}
+			mt.store[key] = si
+			size += uint64(len(r.MetricName)) + storeOverhead
+			cnt++
+		}
+		if err := zr.Close(); err != nil {
+			return nil, fmt.Errorf("cannot close gzip reader: %w", err)
+		}
+
+		mt.currentSizeBytes.Store(size)
+		mt.currentItemsCount.Store(cnt)
+		logger.Infof("loaded state from disk, records: %d, total size: %d", cnt, size)
+	}
+	return mt, nil
+}
+
+func (mt *Tracker) nextRecordLocked() *statItem {
+	n := len(mt.statItemBuf) + 1
+	if n <= cap(mt.statItemBuf) {
+		mt.statItemBuf = mt.statItemBuf[:n]
+		st := &mt.statItemBuf[n-1]
+		return st
+	}
+	// allocate a new slice instead of reallocating exist
+	// it saves memory and reduces GC pressure
+	siStoreNew := make([]statItem, 0, statItemBufSize)
+	mt.statItemBuf = siStoreNew[:1]
```

Comment:
why not ` siStoreNew[:0]`?

Using a sync.pool seems simpler than managing an internal buffer. It also allows to register metrics concurrently. The numbers of allocations will be bigger with sync.pool, sure, but it doesn't seem critical as churn in metric names is expected to be very low.
What are the reasons to go with this approach?

Reply:
The main reason is reduce GC pressure and memory usage.

It's not possible to use sync.Pool anyhow for this case. It's designed for temporary allocations of objects. Those allocations are long-lived.

 The main "trick" of this code is to make batch allocations on heap. Instead of allocating 1024 new objects for each item, this buffer allocates slice for 1024 objects and GC tracks only this slice.

Reply:
Ah, so `bytesutil.ToUnsafeString` will return a string that still points to underlying array where it resides. 
So in result you have N strings pointing to the same array. Once array capacity is reached, you allocate a new one and new strings start pointing on it.
Since old strings still point to a prev arrays - those arrays arent GCed. 
That's clever. I never seen this before.

Reply:
I'm not sure this level of optimization is actually needed here, though. My understanding was that number of unique metric names if pretty limited, thousands at max. And just simply `string()` allocating them shouldn't be an issue.

Reply:
In my local test with 2M tracked metric names, GC time increased up to 4 seconds ( while usually it takes 20ms).

Reply:
Yeah, i don't think we'll ever observe 2M in the wild. I'm not against this optimization, though. I'd only ask to add a comment explaining why you're doing it. 

Maybe @valyala should also take a look.

Reply:
Let's proceed without @valyala. We can always change the internal implementation later.

Reply:
I think it's needed. Because cache limit could be reached with time and it could confuse users. At first glance it will look like high CPU usage and it could be hard to debug.

I added comments to the separate PR at master branch. 

 I'm going to back-port changes from it to this version as soon as it will be approved.